microprofile/WFLY-12758_integrate_opentracing.adoc

```diff
@@ -0,0 +1,111 @@
+
+=== Nice-to-Have Requirements
+
+N/A
+
+=== Non-Requirements
+
+Other tracer implementations will not be configureable from the subsystem management API.
+
+== Implementation Plan
+
+We will only cover the creation of a Jaeger tracer configuration. Other tracer implementations may come afterwards but are not currently 
+
+* Create a `jaeger-tracer` resource that define the configuration of a Jaeger tracer. This resource will expose all Jaeger Tracer configuration attributes:
+    ** `propagation`: The supported trace context propagation formats:
+        *** JAEGER: The default Jaeger trace context propagation format.
+        *** B3: The Zipkin B3 trace context propagation format.
+    ** `sampler-type`: The type of sampler to use in the tracer. Optional. Valid values: remote (default),ratelimiting, probabilistic, const.
+    ** `sampler-param`: The floating point value that makes sense for the correct samplerType.
+    ** `sampler-manager-host-port`: the port of the sampling manager that can provide sampling strategy to this service.
+    ** `sender-binding`: The outbound socket binding to connecto the Agent.
+    ** `sender-endpoint`: The endpoint, like https://jaeger-collector:14268/api/traces
+    ** `sender-user`: The Basic Auth username to be added on Authorization headers for requests sent to the endpoint.
+    ** `sender-auth-password`: The Basic Auth password to be added on Authorization headers for requests sent to the endpoint.
+    ** `sender-auth-token`: The Auth Token to be added as "Bearer" on Authorization headers for requests sent to the endpoint.
+    ** `reporter-log-spans`: Boolean to indicates whether the reporter should log the spans.
+    ** `reporter-flush-interval`: The flush interval when reporting spans remotely in milliseconds.
+    ** `reporter-max-queue-size`: The reporter's maximum queue size.
+    ** `tracer_id_128bit`: Opt-in to use 128 bit traceIds. By default, uses 64 bits.
+    ** `tracer-tags`: A comma separated list of name = value tracer level tags, which get added to all reported spans. The value can also refer to an environment variable using the format ${envVarName:default}, where the :default is optional, and identifies a value to be used if the environment variable cannot be found.
+
+* Create a `default-tracer` attribute on the `microprofile-opentracing-smallrye` resource to define which tracer configuration is the default one to inject if none is specified in the deployment.
+
+* During the processing of the deployment, the module required for the tracer implementations will be added to the classloader as optional dependencies. Then the web.xml will be parsed and if an initial context parameter `smallrye.opentracing.tracer.configuration` is found then the configuration of the matching name will be used to create the tracer, otherwise the default configuration will be set as `smallrye.opentracing.tracer.configuration` value.
+
```

Comment:
Lets assume WF server defines 3 Jeager configurations, first of them is marked as default, server running in production.

case a) Admin comes and marks second one as a default. What happens with deployments which do not explicitly define tracer configuration ?

case b) Admin comes and marks second one as a default and removes first one. What happens with deployments which do not explicitly define tracer configuration? Can be first one removed ?

case c) Admin comes and removes first one. Can be first one removed ? Which config becomes the default one ? What happens with deployments which do not explicitly define tracer configuration?

Reply:
case a) nothing if those deployments are running, they still use old configuration until redeployment / reload
case b) nothing if those deployments are running, they still use old configuration until redeployment / reload.  We should be able to remove the configuration since the service would still be there.
case c) You need to undefine default when removing first one. No default is supported, they would get no configuration and thus rely on some external mean of configuration (env variables for example) or use a NoopTracer (as per specs).

Reply:
> they still use old configuration until redeployment / reload

d) Is there a way to get information which configuration is in reality used by deployments?
Motivation is to detect such orphans which are using non-existing / non-relevant settings.

Based from your feedback I am thinking about new scenarios - reconfiguration.

e) When admin performs changes to the config (e.g. sampling rate change) are these changes propagated to all the relevant deployments or reload is needed ?

f) Will any change to configuration return "reload required" ?
This would ensure consistent state between server configuration for jaeger vs. what jaeger config is used in apps

Reply:
d) Well it is either specified in the deployment via an init-parameter or it is injected. We could add a deployment subsystem for this
e) reload is required
f) yes I should add this to the doc

Reply:
d) is about deployments which do not explicitly define tracer configuration, follow up question for a) and b)

Reply:
Ok, well the trick is the tracer is created by the deployment 'itself' via a ServletContextListener. You would want to have some kind of way to know the source of the tracer configuration in the deployment model ?

Reply:
I want to know if there is / will be a way to know which configuration is used by deployments including all the parameters.
Asking because the configuration can change in the management model, but if "reload required" will be asked then config in management model and config used by applications will be in sync.

Reply:
d) ok we could add the name to the deployment subsystem so that the used tracer configuration is known.

Reply:
BUT there is the case of a tracer resolver. This is done via ServiceLoader by the deployment itself

Reply:
Would something like this suits you ? "microprofile-opentracing-smallrye" => {
                "tracer-configuration" => {
                    "sampler-type" => "const",
                    "sampler-param" => 1.0,
                    "propagation" => undefined,
                    "sampler-manager-host-port" => undefined,
                    "sender-binding" => undefined,
                    "sender-endpoint" => undefined,
                    "sender-auth-token" => undefined,
                    "sender-auth-user" => undefined,
                    "sender-auth-password" => undefined,
                    "reporter-log-spans" => undefined,
                    "reporter-flush-interval" => undefined,
                    "reporter-max-queue-size" => undefined,
                    "tags" => undefined,
                    "tracer_id_128bit" => undefined
                },
                "tracer-configuration-name" => "jaeger"
            }

Reply:
Looks good